README.md
![Maintenance](https://img.shields.io/badge/maintenance-experimental-blue.svg)
[![crates.io](https://img.shields.io/crates/v/libftd2xx-cc1101.svg)](https://crates.io/crates/libftd2xx-cc1101)
[![docs.rs](https://docs.rs/libftd2xx-cc1101/badge.svg)](https://docs.rs/libftd2xx-cc1101/)
[![CI](https://github.com/CirrusNeptune/libftd2xx-cc1101/workflows/CI/badge.svg)](https://github.com/CirrusNeptune/libftd2xx-cc1101/actions)

# libftd2xx-cc1101

I/O library for using [CC1101] RF transceiver via [libftd2xx-rs].

## Usage
Simply add this crate as a dependency in your `Cargo.toml`.

```toml
[dependencies]
libftd2xx-cc1101 = "~0.1.0"
```

Also follow the README guidance in [libftd2xx-rs] to set up the connection with the FTDI device.

[RX] and [TX] examples are good starting points.

## References

* [CC1101 Datasheet]

[CC1101]: https://www.ti.com/product/CC1101
[libftd2xx-rs]: https://github.com/newAM/libftd2xx-rs
[RX]: https://github.com/CirrusNeptune/libftd2xx-cc1101/blob/main/examples/rx.rs
[TX]: https://github.com/CirrusNeptune/libftd2xx-cc1101/blob/main/examples/tx.rs
[CC1101 Datasheet]: https://www.ti.com/lit/gpn/cc1101


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2021-06-11
### Added
- Initial release

[0.1.0]: https://github.com/CirrusNeptune/libftd2xx-cc1101/releases/tag/0.1.0


tests/mpsse-macro.rs
#![recursion_limit = "1024"]
//#![feature(trace_macros)]
//trace_macros!(true);

use libftd2xx::{
    ClockBits, ClockBitsIn, ClockBitsOut, ClockData, ClockDataIn, ClockDataOut, MpsseCmd,
    TimeoutError,
};
use libftd2xx_cc1101::regs::*;
use libftd2xx_cc1101::*;
use std::convert::TryInto;

struct MockFtdi {
    pub write_data: Vec<u8>,
    pub read_len: usize,
}

impl MockFtdi {
    fn new() -> Self {
        Self {
            write_data: Vec::new(),
            read_len: 0,
        }
    }

    fn write_all(&mut self, buf: &[u8]) -> Result<(), TimeoutError> {
        self.write_data.extend_from_slice(&buf);
        Ok(())
    }

    fn read_all(&mut self, buf: &mut [u8]) -> Result<(), TimeoutError> {
        self.read_len += buf.len();
        Ok(())
    }
}

#[test]
fn commands() {
    let mut ftdi = MockFtdi::new();

    mpsse! {
        ftdi <=> {
            sres();
            sidle();
            write_tx_fifo([11, 22, 33, 44]);
            write_pa_table([55, 66, 77, 88]);
            stx();
            set_iocfg2(IOCFG2::new());
            set_freq(123456);
            set_sync(512);
            write_all_regs(Regs::new());
        };
    }

    assert_eq!(
        ftdi.write_data,
        [
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            Command::SRES as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            Command::SIDLE as u8,
            ClockDataOut::MsbNeg as u8,
            0x4 as u8,
            0x0 as u8,
            0x7f as u8,
            11 as u8,
            22 as u8,
            33 as u8,
            44 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockDataOut::MsbNeg as u8,
            0x4 as u8,
            0x0 as u8,
            0x7e as u8,
            55 as u8,
            66 as u8,
            77 as u8,
            88 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            Command::STX as u8,
            ClockDataOut::MsbNeg as u8,
            0x1 as u8,
            0x0 as u8,
            RegAddrs::IOCFG2 as u8,
            0x0 as u8,
            ClockDataOut::MsbNeg as u8,
            0x5 as u8,
            0x0 as u8,
            RegAddrs::FREQ as u8,
            1 as u8,
            RegAddrs::FREQ as u8 + 1,
            226 as u8,
            RegAddrs::FREQ as u8 + 2,
            64 as u8,
            ClockDataOut::MsbNeg as u8,
            0x3 as u8,
            0x0 as u8,
            RegAddrs::SYNC as u8,
            2 as u8,
            RegAddrs::SYNC as u8 + 1,
            0 as u8,
            ClockDataOut::MsbNeg as u8,
            47 as u8,
            0x0 as u8,
            0x40 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
        ]
    );
    assert_eq!(ftdi.read_len, 0);
}

#[test]
fn command_reads() {
    let mut ftdi = MockFtdi::new();

    mpsse! {
        ftdi <=> {
            let _sres_status = sres();
            let _sidle_status = sidle();
            let _write_tx_status = write_tx_fifo([11, 22, 33, 44]);
            let _stx_status = stx();
            let _read_data_only = read_rx_fifo(4);
            let (_read_data_status, _read_data) = read_rx_fifo(4);
            let _read_pa_only = read_pa_table(4);
            let (_read_pa_status, _read_pa) = read_pa_table(4);
            let _lqi = lqi();
            let (_lqi_status, _lqi2) = lqi();
            let _partnum = partnum();
            let _wortime = wortime();
            let (_wortime_status, _wortime2) = wortime();
            let _iocfg2 = iocfg2();
            let _iocfg2_status = set_iocfg2(IOCFG2::new());
            let _freq = freq();
            let _freq_status = set_freq(123456);
            let _sync = sync();
            let _sync_status = set_sync(512);
            let _regs = read_all_regs();
            let (_regs_status, _regs2) = read_all_regs();
            let _regs_status2 = write_all_regs(Regs::new());
        };
    }

    assert_eq!(
        ftdi.write_data,
        [
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            Command::SRES as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            Command::SIDLE as u8,
            ClockData::MsbPosIn as u8,
            0x4 as u8,
            0x0 as u8,
            0x7f as u8,
            11 as u8,
            22 as u8,
            33 as u8,
            44 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            Command::STX as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            0xff as u8,
            ClockDataIn::MsbNeg as u8,
            0x3 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            0xff as u8,
            ClockDataIn::MsbNeg as u8,
            0x3 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            0xfe as u8,
            ClockDataIn::MsbNeg as u8,
            0x3 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            0xfe as u8,
            ClockDataIn::MsbNeg as u8,
            0x3 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            ReadRegAddrs::LQI as u8,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            ReadRegAddrs::LQI as u8,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            ReadRegAddrs::PARTNUM as u8,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            ReadRegAddrs::WORTIME as u8,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            ReadRegAddrs::WORTIME as u8 + 1,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            ReadRegAddrs::WORTIME as u8,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            ReadRegAddrs::WORTIME as u8 + 1,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::IOCFG2 as u8 + 0x80,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockData::MsbPosIn as u8,
            0x1 as u8,
            0x0 as u8,
            RegAddrs::IOCFG2 as u8,
            0x0 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::FREQ as u8 + 0x80,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::FREQ as u8 + 0x80 + 1,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::FREQ as u8 + 0x80 + 2,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockData::MsbPosIn as u8,
            0x5 as u8,
            0x0 as u8,
            RegAddrs::FREQ as u8,
            1 as u8,
            RegAddrs::FREQ as u8 + 1,
            226 as u8,
            RegAddrs::FREQ as u8 + 2,
            64 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::SYNC as u8 + 0x80,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::SYNC as u8 + 0x80 + 1,
            ClockBitsIn::MsbNeg as u8,
            0x7 as u8,
            ClockData::MsbPosIn as u8,
            0x3 as u8,
            0x0 as u8,
            RegAddrs::SYNC as u8,
            2 as u8,
            RegAddrs::SYNC as u8 + 1,
            0 as u8,
            ClockBitsOut::MsbNeg as u8,
            0x7 as u8,
            RegAddrs::IOCFG2 as u8 + 0xC0,
            ClockDataIn::MsbNeg as u8,
            46 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockBits::MsbPosIn as u8,
            0x7 as u8,
            RegAddrs::IOCFG2 as u8 + 0xC0,
            ClockDataIn::MsbNeg as u8,
            46 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x0 as u8,
            0xb as u8,
            ClockData::MsbPosIn as u8,
            47 as u8,
            0x0 as u8,
            0x40 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            0x0 as u8,
            MpsseCmd::SetDataBitsLowbyte as u8,
            0x8 as u8,
            0xb as u8,
            MpsseCmd::SendImmediate as u8,
        ]
    );
    assert_eq!(ftdi.read_len, 197);
}


examples/rx.rs
use libftd2xx::{Ft232h, FtStatus, Ftdi, FtdiMpsse, MpsseSettings, TimeoutError};
use libftd2xx_cc1101::regs::{ModFormat, SyncMode};
use libftd2xx_cc1101::CC1101;
use std::convert::TryInto;
use std::io::Read;
use std::time::Duration;

fn initialize_mpsse<Ft: FtdiMpsse>(ftdi: &mut Ft) -> Result<(), TimeoutError> {
    const SETTINGS: MpsseSettings = MpsseSettings {
        reset: true,
        in_transfer_size: 4096,
        read_timeout: Duration::from_secs(1),
        write_timeout: Duration::from_secs(1),
        latency_timer: Duration::from_millis(16),
        mask: 0xb,
        clock_frequency: Some(1_000_000),
    };

    // Attempt initialize_mpsse up to 3 times
    let mut err = Err(TimeoutError::FtStatus(FtStatus::OTHER_ERROR));
    for _ in 0..3 {
        err = ftdi.initialize_mpsse(&SETTINGS);
        if err.is_ok() {
            return Ok(());
        }
    }
    err
}

fn main() {
    let ft = Ftdi::new().expect("unable to Ftdi::new");
    let mut ftdi: Ft232h = ft.try_into().expect("not a Ft232h");

    initialize_mpsse(&mut ftdi).expect("unable to initialize mpsse");

    let mut cc1101 = CC1101::new(&mut ftdi);
    cc1101
        .initialize(|regs, pa_table| {
            regs.set_freq(1093805); // 433.943634 MHz
            regs.set_mdmcfg3(regs.mdmcfg3().with_drate_m(0x67)); // 2.2254 kBaud
            regs.set_mdmcfg2(
                regs.mdmcfg2()
                    .with_mod_format(ModFormat::AskOok)
                    .with_sync_mode(SyncMode::None),
            );
            *pa_table = [0, 0x60, 0, 0, 0, 0, 0, 0]; // 0dBm OOK
        })
        .expect("unable to initialize cc1101");

    let mut cc1101_reader = cc1101.reader::<32>();
    let mut buf = [0_u8; 32];
    cc1101_reader
        .read_exact(&mut buf)
        .expect("unable to read_exact");
    println!("{:?}", buf);
}


examples/tx.rs
use libftd2xx::{Ft232h, FtStatus, Ftdi, FtdiMpsse, MpsseSettings, TimeoutError};
use libftd2xx_cc1101::regs::{ModFormat, SyncMode};
use libftd2xx_cc1101::CC1101;
use std::convert::TryInto;
use std::io::Write;
use std::time::Duration;

fn initialize_mpsse<Ft: FtdiMpsse>(ftdi: &mut Ft) -> Result<(), TimeoutError> {
    const SETTINGS: MpsseSettings = MpsseSettings {
        reset: true,
        in_transfer_size: 4096,
        read_timeout: Duration::from_secs(1),
        write_timeout: Duration::from_secs(1),
        latency_timer: Duration::from_millis(16),
        mask: 0xb,
        clock_frequency: Some(1_000_000),
    };

    // Attempt initialize_mpsse up to 3 times
    let mut err = Err(TimeoutError::FtStatus(FtStatus::OTHER_ERROR));
    for _ in 0..3 {
        err = ftdi.initialize_mpsse(&SETTINGS);
        if err.is_ok() {
            return Ok(());
        }
    }
    err
}

fn main() {
    let ft = Ftdi::new().expect("unable to Ftdi::new");
    let mut ftdi: Ft232h = ft.try_into().expect("not a Ft232h");

    initialize_mpsse(&mut ftdi).expect("unable to initialize mpsse");

    let mut cc1101 = CC1101::new(&mut ftdi);
    cc1101
        .initialize(|regs, pa_table| {
            regs.set_freq(1093805); // 433.943634 MHz
            regs.set_mdmcfg3(regs.mdmcfg3().with_drate_m(0x67)); // 2.2254 kBaud
            regs.set_mdmcfg2(
                regs.mdmcfg2()
                    .with_mod_format(ModFormat::AskOok)
                    .with_sync_mode(SyncMode::None),
            );
            *pa_table = [0, 0x60, 0, 0, 0, 0, 0, 0]; // 0dBm OOK
        })
        .expect("unable to initialize cc1101");

    let mut cc1101_writer = cc1101.writer::<32>();
    cc1101_writer
        .write_all(&[11, 22, 33, 44])
        .expect("unable to write_all");
    cc1101_writer.flush().expect("unable to flush");
}


src/io.rs
//! This module contains TX/RX FIFO implementations conforming to [`std::io`] traits.

#![deny(missing_docs, unsafe_code, warnings)]

use crate::regs::{FifoThreshold, GDOCfg, RXBYTES, TXBYTES};
use crate::{Command, MpsseCmdBuilder, ReadRegAddrs, RegAddrs, Status, CC1101};
use libftd2xx::{
    mpsse, ClockBits, ClockBitsIn, ClockBitsOut, ClockDataIn, FtdiCommon, TimeoutError,
};
use log::trace;
use ringbuffer::{ConstGenericRingBuffer, RingBuffer, RingBufferRead, RingBufferWrite};
use std::io;

const fn size_to_rx_threshold(size: usize) -> FifoThreshold {
    match size {
        1..=4 => FifoThreshold::Tx61Rx4,
        5..=8 => FifoThreshold::Tx57Rx8,
        9..=12 => FifoThreshold::Tx53Rx12,
        13..=16 => FifoThreshold::Tx49Rx16,
        17..=20 => FifoThreshold::Tx45Rx20,
        21..=24 => FifoThreshold::Tx41Rx24,
        25..=28 => FifoThreshold::Tx37Rx28,
        29..=32 => FifoThreshold::Tx33Rx32,
        33..=36 => FifoThreshold::Tx29Rx36,
        37..=40 => FifoThreshold::Tx25Rx40,
        41..=44 => FifoThreshold::Tx21Rx44,
        45..=48 => FifoThreshold::Tx17Rx48,
        49..=52 => FifoThreshold::Tx13Rx52,
        53..=56 => FifoThreshold::Tx9Rx56,
        57..=60 => FifoThreshold::Tx5Rx60,
        61..=64 => FifoThreshold::Tx1Rx64,
        _ => FifoThreshold::Tx61Rx4,
    }
}

const fn size_to_tx_threshold(size: usize) -> FifoThreshold {
    match size {
        57..=60 => FifoThreshold::Tx61Rx4,
        53..=56 => FifoThreshold::Tx57Rx8,
        49..=52 => FifoThreshold::Tx53Rx12,
        45..=48 => FifoThreshold::Tx49Rx16,
        41..=44 => FifoThreshold::Tx45Rx20,
        37..=40 => FifoThreshold::Tx41Rx24,
        33..=36 => FifoThreshold::Tx37Rx28,
        29..=32 => FifoThreshold::Tx33Rx32,
        25..=28 => FifoThreshold::Tx29Rx36,
        21..=24 => FifoThreshold::Tx25Rx40,
        17..=20 => FifoThreshold::Tx21Rx44,
        13..=16 => FifoThreshold::Tx17Rx48,
        9..=12 => FifoThreshold::Tx13Rx52,
        5..=8 => FifoThreshold::Tx9Rx56,
        1..=4 => FifoThreshold::Tx5Rx60,
        0 => FifoThreshold::Tx1Rx64,
        _ => FifoThreshold::Tx61Rx4,
    }
}

fn timeout_error(_: TimeoutError) -> io::Error {
    io::Error::from(io::ErrorKind::TimedOut)
}

/// Buffered RX-FIFO reader for CC1101 that implements [`io::Read`] trait.
///
/// Instances are created via [`CC1101::reader`].
///
/// Maintains a circular buffer of user-configurable size via the `BUF_CAP` parameter.
/// The buffer size should be adjusted for the application to balance between low access latency
/// and low frequency of underlying read operations.
pub struct FifoReader<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> {
    cc1101: &'c mut CC1101<'f, Ft>,
    rb: ConstGenericRingBuffer<u8, BUF_CAP>,
}

impl<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> FifoReader<'f, 'c, Ft, BUF_CAP> {
    pub(crate) fn new(cc1101: &'c mut CC1101<'f, Ft>) -> Self {
        assert!(BUF_CAP > 0 && BUF_CAP <= 64);
        Self {
            cc1101,
            rb: ConstGenericRingBuffer::new(),
        }
    }

    fn srx(&mut self) -> io::Result<usize> {
        let cur_state = self.cc1101.get_state().map_err(timeout_error)?;
        self.cc1101.flush_fifos(cur_state).map_err(timeout_error)?;

        mpsse! {
            const (SRX_RXBYTES, READ_LEN) = {
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, Command::SRX as u8, 8);
                clock_bits_out(ClockBitsOut::MsbNeg, ReadRegAddrs::RXBYTES as u8, 8);
                clock_bits_in(ClockBitsIn::MsbNeg, 8);
                set_gpio_lower(0x8, 0xb);
                send_immediate();
            };
        }

        trace!(">CC1101 SRX/RXBYTES");
        self.cc1101
            .ftdi
            .write_all(&SRX_RXBYTES)
            .map_err(timeout_error)?;
        let mut buf = [0_u8; READ_LEN];
        self.cc1101.ftdi.read_all(&mut buf).map_err(timeout_error)?;
        let rxbytes = RXBYTES::from_bytes(buf).num_rxbytes() as usize;
        trace!("<CC1101 SRX/RXBYTES {}", rxbytes);

        Ok(rxbytes)
    }

    fn wait_for_rx_fifo_len(&mut self, len: usize) -> io::Result<()> {
        let new_iocfg2 = self
            .cc1101
            .regs
            .iocfg2()
            .with_gdo2_cfg(GDOCfg::RxFifoThreshold);
        let new_fifothr = self
            .cc1101
            .regs
            .fifothr()
            .with_fifo_thr(size_to_rx_threshold(len));

        mpsse! {
            let (wait, READ_LEN) = {
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, RegAddrs::IOCFG2 as u8, 8);
                clock_bits_out(ClockBitsOut::MsbNeg, new_iocfg2.into_bytes()[0], 8);
                clock_bits_out(ClockBitsOut::MsbNeg, RegAddrs::FIFOTHR as u8, 8);
                clock_bits_out(ClockBitsOut::MsbNeg, new_fifothr.into_bytes()[0], 8);
                set_gpio_lower(0x8, 0xb);
                wait_on_io_high();
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, ReadRegAddrs::RXBYTES as u8, 8);
                clock_bits_in(ClockBitsIn::MsbNeg, 8);
                set_gpio_lower(0x8, 0xb);
            };
        }

        trace!(">CC1101 WAITING_FOR_RXBYTES >= {}", len);
        self.cc1101.ftdi.write_all(&wait).map_err(timeout_error)?;
        let mut buf = [0_u8; READ_LEN];
        while self.cc1101.ftdi.read_all(&mut buf).is_err() {}
        let rxbytes = RXBYTES::from_bytes(buf).num_rxbytes() as usize;
        trace!("<CC1101 WAITING_FOR_RXBYTES DONE {}", rxbytes);

        if rxbytes >= len {
            Ok(())
        } else {
            Err(io::Error::new(
                io::ErrorKind::Other,
                "WAITING_FOR_RXBYTES failed",
            ))
        }
    }

    fn read_into_buffer(&mut self, len: usize) -> io::Result<()> {
        mpsse! {
            let read_rx_fifo = {
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, 0xff, 8);
                clock_data_in(ClockDataIn::MsbNeg, len);
                set_gpio_lower(0x8, 0xb);
                send_immediate();
            };
        }

        trace!(">CC1101 RXFIFO {}", len);
        self.cc1101
            .ftdi
            .write_all(&read_rx_fifo)
            .map_err(timeout_error)?;
        let mut buf = [0_u8; BUF_CAP];
        self.cc1101
            .ftdi
            .read_all(&mut buf[0..len])
            .map_err(timeout_error)?;
        trace!("<CC1101 RXFIFO {:x?}", &buf[0..len]);

        for b in buf[0..len].iter() {
            self.rb.push(*b);
        }

        Ok(())
    }

    fn fill_buffer(&mut self) -> io::Result<()> {
        let rxbytes = self.srx()?;

        let read_len = self.rb.capacity() - self.rb.len();
        if rxbytes < read_len {
            self.wait_for_rx_fifo_len(read_len)?;
        }

        self.read_into_buffer(read_len)
    }
}

impl<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> io::Read for FifoReader<'f, 'c, Ft, BUF_CAP> {
    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
        let read_len = std::cmp::min(BUF_CAP, buf.len());

        if read_len > self.rb.len() {
            self.fill_buffer()?;
        }

        for b in buf[0..read_len].iter_mut() {
            *b = self.rb.dequeue().unwrap();
        }

        Ok(read_len)
    }
}

/// Buffered TX-FIFO writer for CC1101 that implements [`io::Write`] trait.
///
/// Instances are created via [`CC1101::writer`].
///
/// Maintains a circular buffer of user-configurable size via the `BUF_CAP` parameter.
/// The buffer size should be adjusted for the application to balance between low access latency
/// and low frequency of underlying write operations.
pub struct FifoWriter<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> {
    cc1101: &'c mut CC1101<'f, Ft>,
    rb: ConstGenericRingBuffer<u8, BUF_CAP>,
}

impl<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> FifoWriter<'f, 'c, Ft, BUF_CAP> {
    pub(crate) fn new(cc1101: &'c mut CC1101<'f, Ft>) -> Self {
        assert!(BUF_CAP > 0 && BUF_CAP <= 64);
        Self {
            cc1101,
            rb: ConstGenericRingBuffer::new(),
        }
    }

    fn txbytes(&mut self) -> io::Result<usize> {
        mpsse! {
            const (TXBYTES, READ_LEN) = {
                set_gpio_lower(0x0, 0xb);
                clock_bits(ClockBits::MsbPosIn, ReadRegAddrs::TXBYTES as u8, 8);
                clock_bits_in(ClockBitsIn::MsbNeg, 8);
                set_gpio_lower(0x8, 0xb);
                send_immediate();
            };
        }

        trace!(">CC1101 TXBYTES");
        self.cc1101
            .ftdi
            .write_all(&TXBYTES)
            .map_err(timeout_error)?;
        let mut buf = [0_u8; READ_LEN];
        self.cc1101.ftdi.read_all(&mut buf).map_err(timeout_error)?;
        let status = Status::from_bytes([buf[0]]);
        let txbytes = TXBYTES::from_bytes([buf[1]]).num_txbytes() as usize;
        trace!("<CC1101 TXBYTES {:?} {}", status, txbytes);

        self.cc1101
            .flush_fifos(status.state())
            .map_err(timeout_error)?;

        Ok(txbytes)
    }

    fn wait_for_tx_fifo_len(&mut self, len: usize) -> io::Result<()> {
        let new_iocfg2 = self
            .cc1101
            .regs
            .iocfg2()
            .with_gdo2_cfg(GDOCfg::TxFifoThreshold);
        let new_fifothr = self
            .cc1101
            .regs
            .fifothr()
            .with_fifo_thr(size_to_tx_threshold(len));

        mpsse! {
            let (wait, READ_LEN) = {
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, RegAddrs::IOCFG2 as u8, 8);
                clock_bits_out(ClockBitsOut::MsbNeg, new_iocfg2.into_bytes()[0], 8);
                clock_bits_out(ClockBitsOut::MsbNeg, RegAddrs::FIFOTHR as u8, 8);
                clock_bits_out(ClockBitsOut::MsbNeg, new_fifothr.into_bytes()[0], 8);
                clock_bits_out(ClockBitsOut::MsbNeg, Command::STX as u8, 8);
                set_gpio_lower(0x8, 0xb);
                wait_on_io_low();
                set_gpio_lower(0x0, 0xb);
                clock_bits_out(ClockBitsOut::MsbNeg, ReadRegAddrs::TXBYTES as u8, 8);
                clock_bits_in(ClockBitsIn::MsbNeg, 8);
                set_gpio_lower(0x8, 0xb);
            };
        }

        trace!(">CC1101 WAITING_FOR_TXBYTES <= {}", len);
        self.cc1101.ftdi.write_all(&wait).map_err(timeout_error)?;
        let mut buf = [0_u8; READ_LEN];
        while self.cc1101.ftdi.read_all(&mut buf).is_err() {}
        let txbytes = TXBYTES::from_bytes(buf).num_txbytes() as usize;
        trace!("<CC1101 WAITING_FOR_TXBYTES DONE {}", txbytes);

        if txbytes <= len {
            Ok(())
        } else {
            Err(io::Error::new(
                io::ErrorKind::Other,
                "WAITING_FOR_TXBYTES failed",
            ))
        }
    }

    fn write_from_buffer(&mut self) -> io::Result<()> {
        let write_len = self.rb.len();
        let mut buf = [0_u8; 65];
        buf[0] = 0x7F;
        for b in buf[1..=write_len].iter_mut() {
            *b = self.rb.dequeue().unwrap();
        }

        trace!(">CC1101 TXFIFO {:x?}", &buf[1..=write_len]);
        self.cc1101
            .ftdi
            .write_all(
                MpsseCmdBuilder::new()
                    .cs_low()
                    .clock_data_out(&buf[0..=write_len])
                    .cs_high()
                    .cs_low()
                    .clock_bits_out(Command::STX as u8, 8)
                    .cs_high()
                    .as_slice(),
            )
            .map_err(timeout_error)
    }

    fn write_cycle(&mut self) -> io::Result<()> {
        let txbytes = self.txbytes()?;

        if 64 - txbytes < self.rb.len() {
            self.wait_for_tx_fifo_len(64 - self.rb.len())?;
        }

        self.write_from_buffer()
    }

    /// Block until all pending bytes have been transmitted.
    pub fn finish(&mut self) -> io::Result<()> {
        if !self.rb.is_empty() {
            self.write_cycle()?;
        }

        let txbytes = self.txbytes()?;

        if txbytes > 0 {
            self.wait_for_tx_fifo_len(0)?;
        }

        Ok(())
    }
}

impl<'f, 'c, Ft: FtdiCommon, const BUF_CAP: usize> io::Write for FifoWriter<'f, 'c, Ft, BUF_CAP> {
    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
        if self.rb.is_full() {
            self.flush()?;
        }

        let write_len = std::cmp::min(self.rb.capacity() - self.rb.len(), buf.len());
        for b in buf[0..write_len].iter() {
            self.rb.push(*b);
        }

        Ok(write_len)
    }

    fn flush(&mut self) -> io::Result<()> {
        self.write_cycle()
    }
}
